`dotnet tool install --local` works by default, by creating a manifest by default, utilizing `--create-manifest-if-needed`

nagilson opened the pull request:
Resolves https://github.com/dotnet/sdk/issues/15254#issuecomment-1470759579
Resolves part of https://github.com/dotnet/sdk/issues/36830 by fixing the default, but does not fix `-d`. That should be done in another PR.

# What's this do

Makes `createManifestIfNeeded` be enabled by default. This is a breaking change, since users may have relied on the failure behavior before to check if they needed to create a manifest, but now we will create one according to the rules followed by the flag, documented at: https://github.com/dotnet/docs/issues/39600 regardless of if they set the flag or not.

They can turn off the behavior by setting `--create-manifest-if-needed=false`...

Note: We should have someone who knows dotnet tools better review this PR, since I don't really know what side effects there are to making the tool manifest, besides that it puts a file on disk that allows you to run the tool locally. But I think that's what the user wants to do anyway.

* This points to MAIN, please lmk if we want this sooner.

# Why / Should We 

See the comments in the above PR...
The original maintainers were concerned this would cause people to start making tool manifests everywhere and get confused why did the tool not work from everywhere... However, we resolve this concern by only putting the tool manifest at the repo root when possible, and only in the cwd when we can't find a root. It is a strong ux improvement as before tool install local for the first time will always fail.

# Explanation

I reviewed the previous PR https://github.com/dotnet/sdk/pull/31231 and made sure I only impacted the related code added there, and that the code added there worked.
It basically looks for every manifest file, and returns if the manifest logic that already existed found a manifest.
If not:
It looks over the directory until it's the root directory, and searches correctly for .git, then .sln/.slnx (good change), to find a 'repo root'. If it finds one it uses that to make a manifest, otherwise it uses the CWD/probeStart which was existing logic.
It's a bit confusing due to the use of .Value.Value, but apparently the value of the directory path, probePath, is not a string but {""}, so you need to call value twice.

# Code changes

The change simply makes the default factory constructor for the manifest if needed to set it to true. Added tests and modified existing tests. The manifest flag is only utilized by the local tool install command and then the default tool command parser. This does make all tool commands have --create-manifest-if-needed as true, but only the tool install local command does anything with it, so this shouldnt change the outcome of the other commands to my knowledge.

nagilson committed:
Set the default value of Create Manifest If Needed to True

This should enable the findFirst function to create a tool manifest by default. Will break some tests. On initial glance,, the arg behavior only seems to be inherited  by tool install local, which is all I want to impact right now. Although it also gets inherited by the construct command in command parser.

nagilson committed:
Add test to show it uses manifest if needed by default

nagilson committed:
Update Tests

- We dont need two tests showing it should throw, and then it should throw with a message. Kept only the 2nd test.

- Modified the we should throw test to only expect throw if create manifest if needed is false since the default is now true.

nagilson committed:
Fix Test

nagilson committed:
Accept 'false' instead of just empty arity and add 0 arity test

nagilson committed:
Update completions test bash

nagilson commented:
@baronfel Is there a quick way to update the completions tests for each shell to be the current behavior? I think the Arity change here caused the test to break. I was able to update the snapshot by manually running it and copying over the new snapshot for bash, though I'm hoping there's an automated way to avoid having to do that for Fish, Zsh, and others

nagilson committed:
Merge branch 'main' into nagilson-create-tool-manifest-by-default

nagilson committed:
Update snapshots

1. Run Test
2.  dotnet restore .\dotnet.Tests\ /t:CompareCliSnapshots in test folder on test prj
3. dotnet restore .\dotnet.Tests\ /t:UpdateCliSnapshots

nagilson committed:
Merge branch 'nagilson-create-tool-manifest-by-default' of https://github.com/nagilson/sdk into nagilson-create-tool-manifest-by-default

nagilson commented:
Oh awesome, thank you 🍰 🐙

dsplaisted commented on test/dotnet.Tests/CommandTests/Tool/Install/ToolInstallLocalCommandTests.cs:
```diff
@@ -128,22 +128,41 @@ public void WhenRunWithPackageIdItShouldSaveToCacheAndAddToManifestFile()
-            var toolInstallLocalCommand = GetDefaultTestToolInstallLocalCommand();
+            ParseResult parseResult =
+            Parser.Instance.Parse(
+               $"dotnet tool install {_packageIdA.ToString()} --create-manifest-if-needed false");
```
Should we update the text for ToolInstallNoManifestGuide?

> If you intended to install a global tool, add `--global` to the command.
> If you would like to create a manifest, use the `--create-manifest-if-needed` flag with the `dotnet tool install` command, or use `dotnet new tool-manifest`, usually in the repo root directory.

nagilson replied:
That's a thoughtful question and I think worth considering. We could remove the --create-manifest-if-needed part, though I think that obfuscates the fact that there is a manifest as a concept when people hit this error. There may be a lot of users who aren't aware of don't care about the manifest, IMO, so maybe it's good to keep this message as is. It is not misleading/incorrect despite this. I'm also not sure when this will be shown now, besides when using false for create manifest if needed in the old scenarios.

dsplaisted reviewed (commented):
The code change looks good, but I'm wondering if we may want to tweak the design.

Do we actually want to create the manifest by default if you don't specify `--local`?  I think it makes sense to create the manifest by default if someone specified `dotnet tool install --local MyTool`, but if they specify `dotnet tool install MyTool` maybe they wanted a global tool but forgot to specify.

What do you think @baronfel?

nagilson committed:
Remove `ToolInstallNoManifestGuide` and Fix `CannotFindAManifestFile `

`CannotFindAManifestFile` now always provides the list of searched locations.
Remove the string that is now longer needed now that --create-manifest-if-needed is added by default.

Need to look into `ToolCommonNoManifestGuide` and if we still need that

nagilson committed:
Update xlf

Looks like we can keep the common one since its used in uninstall as well as install, but in cases where you set --create-manifest-if-needed to false, which is a case where you likely did want a global tool.

nagilson committed:
Merge branch 'main' into nagilson-create-tool-manifest-by-default

nagilson commented:
> Yes, I think we should change the message. It looks like first a `ToolManifestCannotBeFoundException` is thrown using `CliStrings.CannotFindAManifestFile` as the message. Then in `ToolInstallLocalCommand.GetManifestFilePath`, the exception is rethrown with the additional `CliCommandStrings.ToolInstallNoManifestGuide` message. I think we can get rid of the rethrow and the `CliCommandStrings.ToolInstallNoManifestGuide` string altogether.
> 
> The text for `CliStrings.CannotFindAManifestFile` is:
> 
> > Cannot find a manifest file.
> > For a list of locations searched, specify the "-d" option before the tool name.
> 
> I think specifying the `-d` option doesn't actually work. So we should either make it work or always list the locations searched (and remove the mention of the `-d` option).

Thanks, I followed this up by removing the `-d` flag and doing this by default, as well as removing the string. I also found `ToolCommonNoManifestGuide` but I think we can keep that and I provided more info in the commit details.

nagilson committed:
update xlf again

nagilson committed:
Remove ListOfSearched

nagilson committed:
Update XLF

nagilson committed:
Fix Test

Just search an arbitrary number of the substring
Not ideal, trying this for now

akoeplinger committed:
Fix variables in restore-toolset.sh/.ps1

The productBuild variable got renamed in https://github.com/dotnet/sdk/pull/49080 which caused errors.

nagilson committed:
Fix test string format

nagilson committed:
Merge branch 'main' into nagilson-create-tool-manifest-by-default

nagilson commented:
/azp run

Forgind commented on src/Cli/dotnet/Commands/Tool/Install/ToolInstallLocalCommand.cs:
```diff
@@ -194,18 +194,8 @@ public int InstallNewTool(FilePath manifestFile, PackageId packageId)
-                ? _toolManifestFinder.FindFirst(_createManifestIfNeeded)
-                : new FilePath(_explicitManifestFile);
-        }
-        catch (ToolManifestCannotBeFoundException e)
```
Why don't we want this to be graceful?

nagilson replied:
It throws `ToolManifestCannotBeFoundException` in `findFirst`, which is a child class of `GracefulException`

Forgind reviewed (commented):
> side effects there are to making the tool manifest

The only thing that comes to mind is that if you specify that it's a root manifest, and we ever see that, we stop probing up. If we put this tools manifest at cwd and specify root: true, that might mean we'd fail to find a tool specified in a higher manifest if present...except that we'd just probed for a higher manifest and didn't find one, so we'd basically have to not have a manifest, call this, then create a manifest somewhere higher and install a local tool using it, and then we'd fail here. I don't think that's worth worrying about.

> searches correctly for .git, then .sln/.slnx

Do we want to be able to support other source control systems? I do think we should permit .slnf as well even though you'd need a .sln somewhere, since you can technically have the .sln somewhere random.

That said, I don't see any of that in this PR, just the string tweaks and changing the default 🙂

nagilson committed:
Merge branch 'main' into nagilson-create-tool-manifest-by-default

nagilson committed:
Fix all of the tests

We don't need a verbose message that's a copy of the non verbose message each time....

nagilson committed:
Merge branch 'main' into nagilson-create-tool-manifest-by-default
